{% extends "base.html" %}
{% block script %}
    {% if show_gene_browser %}
    <script type="text/javascript" src="https://www.ncbi.nlm.nih.gov/projects/sviewer/js/sviewer.js"></script>
    {% endif %}
{% endblock %}

{% block header %}
<div>
<pre>BLAST output file:   {{hea.input}}
Query sequence file: {{hea.query}}
{% if hea.best_matching_model %}
RFAM model with best score to a query sequence   <div class="tooltip">?<pre class="tooltiptext">Infered from query sequence by cmscan program.</pre></div>
Family name: {{hea.best_matching_model['target_name']}}
E-value:     {{hea.best_matching_model['E-value']}}{% endif %}
</pre>
</div>

{% endblock %}

{% block content %}
    {% macro create_onehit(data) %}
    <div class="onehit"
         id="{{data.intid}}onehit"
         data-brna_sequence="{{data.sequence}}"
         data-brna_seqname="{{data.seqname}}"
         data-eval="{{data.eval}}">
        <h3 class="onehit_heading" style="background:{{data.h_color}};">
            Hit: {{data.source_seq_name}}
        </h3>
        <div class="onehit_body">
            <div class="row reports">
                <div class="blast_text repitem">
                    <p class="header-bhname">{{data.blast_hit_name}}</p>
                    <pre class="blasttext">
                        <div class="tooltip blasttooltip">
?<pre class="tooltiptext">
This is BLAST alignment as read from the input file</pre>
                        </div>
{{data.blast_text}}</pre>
                </div>
                <div class="repitem">
                    <label class="repheader"><u>Extension report:</u></label>
                    <table>
                        <tr>
                            <th class="left" scope="row">sequence start:</th>
                            <td class="right" id="{{data.intid}}SeqStart">{{data.ext_start}}</td>
                            <td>
                                <div class="tooltip"> ?
                                    <pre class="tooltiptext">
Start position of the extended sequence in genome.
Start index < end index.</pre>
                                </div>
                            </td>
                        </tr>
                        <tr>
                            <th class="left" scope="row">sequence end:</th>
                            <td class="right" id="{{data.intid}}SeqEnd">{{data.ext_end}}</td>
                            <td>
                                <div class="tooltip"> ?
                                    <pre class="tooltiptext">
End position of the extended sequence in genome.
Start index < end index.</pre>
                                </div>
                            </td>
                        </tr>
                        <tr>
                            <th class="left" scope="row">bit score (CM):</th>
                            <td class="right">{{data.rsearchbitscore}}</td>
                            <td>
                                <div class="tooltip"> ?
                                    <pre class="tooltiptext">
The score for aligning extended sequence to CM model
  (computed by RSEARCH -> default,
  infered from rfam or provided by user)</pre>
                                </div>
                            </td>
                        </tr>
                        <tr>
                            <th class="left" scope="row">Homology estimate:</th>
                            <td class="rigth">{{data.h_estimate}}</td>
                            <td>
                                <div class="tooltip"> ?
                                    <pre class="tooltiptext">
Quick homology estimate:
  Not homologous: bit score < 0
  Homologous: bit score > 20 and bit score > 0.5 * query bit score
  Uncertain otherwise</pre>
                                </div>
                            </td>
                        </tr>
                    </table>
                </div>
                <div class="repitem">
                    <div>
                        <label for="{{data.intid}}SeqCheck" class="repheader"><u>Extended sequence fasta:</u></label>
                        <input type="checkbox" class="individualSequenceCheckbox" id="{{data.intid}}SeqCheck">
                        <div class="tooltip repheader">?
                            <pre class="tooltiptext">
Click checkbox to select multiple seuqences.
Fasta header format:
  UID|accession.versionSTRAND start-end</pre>
                        </div>
                    </div>
                    <textarea id="{{data.intid}}FormSeq" readonly="true" rows="5" cols="65">{{data.formated_seq}}</textarea>
                </div>
            </div>
            <div class="row">
                {% for pic in data.pictures %}
                    <figure class="column" data-brna_secondary_structure="{{pic.secondary_structure}}" id="{{data.intid}}{{pic.picname}}">
                        <a href onclick="mypicload(this)">
                            <img class="rnapic" src="{{pic.pic}}" />
                        </a>
                        <figcaption>
                            <label class="repheader">{{pic.picname}}</label>
                            <input type="checkbox" class="individualStructureCheckbox" id="{{data.intid}}{{data.picname}}StrCheck" data-method="{{pic.picname}}">
                            <div class="tooltip repheader">?
                                <pre class="tooltiptext">
Visualisation of predicted secondary structure.
To save the image:
  Right click on the image -> Save Image as.</pre>
                            </div>
                        </figcaption>
                    </figure>
                {% endfor %}
            </div>

        {% if show_gene_browser %}

            <div id="sv{{data.intid}}" data-sv_params="{{data.seqviewurl}}" class="sv">
                <button class= "seqviewbtn" onclick="viewRegion(this)" title="View selected Regions">Load Sequence viewer</button>
            </div>

        {% endif %}
        {% if len(data.msgs) != 0 %}
            <div class="warnmsgs">
                {% for msg in data.msgs %}
                    <p class="warn">{{msg}}</p>
                {% endfor %}
            </div>
        {% endif %}
        </div>
    </div>
    {% endmacro %}

{% for ii in input_list %}
    {{ create_onehit(ii) }}
{% endfor %}
{% endblock %}

{% block footer %}
<h3 class="onehit_heading">
    {% if len(foo.logdup) != 0 %}
        Command, parameters and program log
    {% else %}
        Command and parameters
    {% endif %}
</h3>
<div class="footerwrap">
    <p class="header-command"><u>executed command:</u></p>
    <textarea readonly="true" rows="3" cols="65">{{foo.command}}</textarea>
    <p>
        <u>date:</u>
        {% if foo.exec_date %}
            {{strftime("%H:%M:%S %d. %m. %Y", foo.exec_date)}}
        {% else %}
            Unknown date.
        {% endif %}
    </p>
    <div>
        {% if len(foo.logdup) != 0 %}
            <p><u>program log:</u></p>
            {% for s in foo.logdup %}
                <p class="warn">{{s}}</p>
            {% endfor %}
        {% endif %}
    </div>
    <pre><u>parameters:</u>
{% for par in foo.parameters %}
    <u>{{par[0]}}:</u> {{par[1]}}
{% endfor %}
    </pre>
</div>
<div class="bottom-space"></div>
{% endblock %}